flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientFactoryTest.java

```diff
@@ -130,6 +130,23 @@ public void run() {
 		}
 	}
 
+	@Test
+	public void testNettyClientConnectRetry() throws Exception {
+		NettyClient nettyClient = mock(NettyClient.class);
```

Comment:
We stopped using mockito for tests (for various of reasons), could you try to provide a proper mock of `NettyClient` class? Maybe if you used `NettyTestUtil` you could just implement `NettyClient#connect` method that just throws an exception on the first invocation, but succeeds (returns real/correct `ChannelFuture`) on the second attempt?

Reply:
I think it will be better. But when I try to implement the TestingNettyClient, I found the SucceededChannelFuture only visible to package io.netty.channel. Thus, I need to create a new class of  SucceededChannelFuture in Flink to return a success ChannelFuture. Do you think it is a right way to go?

Reply:
Yes I noticed that and probably implementing `ChannelFuture` is not worth it, hence I suggested to use `NettyTestUtil`. It looks like you could just init a server (`NettyTestUtil#initServer()`) and `NettyClient` from `#initClient()` and you could use it directly, without a need to implement custom `ChannelFuture`? You can check usages of `NettyTestUtil`, like for example `NettyClientServerSslTest#testClientUntrustedCertificate`.

I might be missing something, but it looks like it could work quite easily.

Reply:
Yes. I created a UnstableNettyClient to simulate the first n-1 connection failure and success in the last try.